# Local Embeddings

We will use HuggingFace to create embeddings locally

## References

- https://docs.llamaindex.ai/en/stable/examples/embeddings/huggingface.html#huggingfaceembedding
- Leaderboard : https://huggingface.co/spaces/mteb/leaderboard
- Explaining leaderboard: https://huggingface.co/blog/mteb

## 1 - Using Llama-Index

In [1]:
from llama_index.embeddings import HuggingFaceEmbedding
embed_model = HuggingFaceEmbedding(model_name="BAAI/bge-small-en-v1.5")
embeddings = embed_model.get_text_embedding("Hello World!")

print(len(embeddings))
print(embeddings[:5])

/Users/dave.nielsen/anaconda3/envs/atlas-3/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
config.json: 100%|██████████| 743/743 [00:00<00:00, 1.84MB/s]
model.safetensors: 100%|██████████| 133M/133M [00:01<00:00, 86.4MB/s] 
tokenizer_config.json: 100%|██████████| 366/366 [00:00<00:00, 2.10MB/s]
vocab.txt: 100%|██████████| 232k/232k [00:00<00:00, 1.88MB/s]
tokenizer.json: 100%|██████████| 711k/711k [00:00<00:00, 3.54MB/s]
special_tokens_map.json: 100%|██████████| 125/125 [00:00<00:00, 1.13MB/s]


384
[-0.0032757113222032785, -0.011690814979374409, 0.04155921936035156, -0.038148146122694016, 0.024183066561818123]


### Let's try a few embedding models

See hugging face embedding models (sentence transformers) here : https://huggingface.co/models?library=sentence-transformers&sort=trending

Here are a select models for comparison.  Taken from leaderboard : https://huggingface.co/spaces/mteb/leaderboard

| model name                              | overall score | model params | model size | embedding length | url                                                            |
|-----------------------------------------|---------------|--------------|------------|------------------|----------------------------------------------------------------|
| intfloat/e5-mistral-7b-instruct         | 66.x          | 7.11 B       | 15 GB      | 4096             | https://huggingface.co/intfloat/e5-mistral-7b-instruct         |
| BAAI/bge-large-en-v1.5                  | 64.x          | 335 M        | 1.34 GB    | 1024             | https://huggingface.co/BAAI/bge-large-en-v1.5                  |
| BAAI/bge-small-en-v1.5                  | 62.x          | 33.5 M       | 133 MB     | 384              | https://huggingface.co/BAAI/bge-small-en-v1.5                  |
| sentence-transformers/all-mpnet-base-v2 | 57.8          |              | 438 MB     | 768              | https://huggingface.co/sentence-transformers/all-mpnet-base-v2 |
| sentence-transformers/all-MiniLM-L12-v2 | 56.x          |              | 134 MB     | 384              | https://huggingface.co/sentence-transformers/all-MiniLM-L12-v2 |
| sentence-transformers/all-MiniLM-L6-v2  | 56.x          |              | 91 MB      | 384              | https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2  |

### Benchmark

In [3]:
embedding_models = [
    'BAAI/bge-large-en-v1.5' ,
    'BAAI/bge-small-en-v1.5' ,
    'sentence-transformers/all-mpnet-base-v2' ,
    'sentence-transformers/all-MiniLM-L12-v2' ,
    'sentence-transformers/all-MiniLM-L6-v2' ,
]

import time
import timeit

for model in embedding_models:
    embed_model = HuggingFaceEmbedding(model_name=model)

    embeddings = embed_model.get_text_embedding("Hello World!")
    print(f'model={model}, embeding_length={len(embeddings):,}')
    %timeit (embed_model.get_text_embedding("Hello World!"))
    print()


model=BAAI/bge-large-en-v1.5, embeding_length=1,024
44.6 ms ± 930 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

model=BAAI/bge-small-en-v1.5, embeding_length=384
23.2 ms ± 380 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

model=sentence-transformers/all-mpnet-base-v2, embeding_length=768
26.1 ms ± 367 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

model=sentence-transformers/all-MiniLM-L12-v2, embeding_length=384
24.6 ms ± 2.55 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

model=sentence-transformers/all-MiniLM-L6-v2, embeding_length=384
13 ms ± 96.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)



## Option 2: Encoding Using Sentence Transformers

In [ ]:
from sentence_transformers import SentenceTransformer
#sentences = ["This is an example sentence", "Each sentence is converted"]

model = SentenceTransformer('sentence-transformers/all-MiniLM-L6-v2')
# embeddings = model.encode(sentences)
embeddings = model.encode('a happy dog!')
print(model)
print (len(embeddings))
print(embeddings[:5])

